### DISCUSSION TITLE
How to cleanup using `van.derive()`?

### DISCUSSION THREAD

**Discussion by Ashvith10** [vanjs-org/van#376] [Q&A] [answered]
So far, the posts I've read across here does not address anything about React or Svelte-equivalent cleanups, so I was wondering if there's a alternative way to achieve this?

I am trying to re-write a React toy-project in VanJS, and I want to be able to call a function to stop a particular event.

**Comment by sirenkovladd** [comment]
Could you share an example of what kind of cleanup you are describing?

**Reply by Ashvith10** [reply]
I've not yet written any code snippet at the moment, but what I am trying to do is make use of the `useEffect()` equivalent in VanJS to create a game-loop, if that makes sense. I want to run a physics simulation so that for the value of n, it will loop a canvas-based physics engine, after which, I want to ensure that I will generate a file containing pre-defined paths for the ball. The project is very poorly documented, but [this](https://github.com/code100x/plinkoo/blob/main/frontend/src/pages/Simulation.tsx) is what I am trying to re-write in VanJS.

**Reply by sirenkovladd** [reply]
vanjs has no way to track component cleaning

you will need to implement this logic yourself, using callbacks when you delete an object

but a simpler and more reliable option is to use [van-element](https://van-element.pages.dev/learn/lifecycle.html#dismount) which is a wrapper for [HTMLElement](https://developer.mozilla.org/en-US/docs/Web/API/Web_components/Using_custom_elements#custom_element_lifecycle_callbacks)

**Reply by Tao-VanJS** [reply]
How about this?

```js
const gameRunning = van.state(true)
let ballManager

van.derive(() => {
  if (gameRunning.val) {
    ballManager = new BallManager(...)
    ...
  } else {
    ballManager.stop()
  }
})
```

You need to set `gameRunning` to `false` at some point of your app.

**Reply by sirenkovladd** [reply]
@Tao-VanJS, yes, this is exactly what I meant by sending signals from the parent element to the child that it should be unmounted

**Reply by Ashvith10** [reply]
@Tao-VanJS @sirenkovladd I am not sure if that would allow me to implement the [follow React snippet](https://github.com/code100x/plinkoo/blob/main/frontend/src/pages/Simulation.tsx#L38-L57) in VanJS, which also makes use of `useRef()`?

**Reply by Tao-VanJS** [reply, marked as answer]
`canvasRef` is just a reference to a DOM element. In VanJS, you can achieve the same with a more straightforward way:

```js
const canvasDom = canvas({width: 800, height: 800})

...
van.derive(() => {
  if (gameRunning.val) {
    ballManager = new BallManager(canvasDom, ...)
    ...
  } else {
    ballManager.stop()
  }
})
...
```

**Reply by Ashvith10** [reply]
@Tao-VanJS is there a simple working project that I can refer? Unfortunately after this discussion, I am still not sure on how to prevent memory leaks after page reload. Take the second snippet for example - since the value of `gameRunning.val` may never be set to false on reload or page-change, it would never be freed, right?

**Reply by Tao-VanJS** [reply]
When the page is reloaded, the memory will be automatically cleaned, so you don't need to worry about page reloads. In your application logic, you need to ensure `ballManager.stop()` is called whenever a new instance is created.

**Reply by Tao-VanJS** [reply]
@Ashvith10, btw: just as what @sirenkovladd mentioned, if you're looking for some managed solution of lifecycle events. Van Element might have what you want: https://van-element.pages.dev/learn/lifecycle.

**Reply by Ashvith10** [reply]
@Tao-VanJS thank you for sharing some pointers - my understanding of mounting and umounting was flawed, mapping them with manually allocated memory.

Looks like all of that complexity can be reduced, and I can track mounting and unmounting by following this [example](https://github.com/vanjs-org/van/discussions/78) - the only thing that didn't make sense to me was the part about using "init" - why not just two states - "mounted" and "unmounted"?

**Reply by Tao-VanJS** [reply]
@Ashvith10, in the particular example you pointed, the stage `unmounted` means the DOM node is permanently deleted from the DOM tree, and you can't bring it back by changing the state `stage` again. But stage `init` means by changing the stage to `mounted`, you will bring up the component.
